nodejs/awsx/cmd/provider/index.ts

```diff
@@ -0,0 +1,142 @@
+// Copyright 2016-2022, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as pulumi from "@pulumi/pulumi";
+import { readFileSync } from "fs";
+import { Trail } from "../../cloudtrail";
+import {trailProviderFactory} from "./trail";
+
+class Provider implements pulumi.provider.Provider {
+    // A map of types to provider factories. Calling a factory may return a new instance each
+    // time or return the same provider instance.
+    private readonly typeToProviderFactoryMap: Record<string, () => pulumi.provider.Provider> = {
+        "awsx:cloudtrail:Trail": trailProviderFactory,
+    };
+
+    constructor(readonly version: string, readonly schema: string) {
+        // Register any resources that can come back as resource references that need to be rehydrated.
+        pulumi.runtime.registerResourceModule("eks", "index", {
+            version: version,
+            construct: (name, type, urn) => {
+                switch (type) {
+                    case "awsx:cloudtrail:Trail":
+                        return new Trail(name, {}, { urn });
+                    default:
+                        throw new Error(`unknown resource type ${type}`);
+                }
+            },
+        });
+    }
+
+    async call(token: string, inputs: pulumi.Inputs): Promise<pulumi.provider.InvokeResult> {
+        throw new Error(`unknown method ${token}`);
+    }
+
+    check(urn: pulumi.URN, olds: any, news: any): Promise<pulumi.provider.CheckResult> {
```

Comment:
The reason this has all the provider methods (e.g. check, diff, create, read, update, delete, etc.) in EKS is because EKS used to have some dynamic provider resources which don't work from MLCs, so they were changed to be implemented as regular custom resources in the EKS provider using these methods. But it doesn't look like awsx has any dynamic provider resources, so I suspect you can clean this up considerably here, deleting all the methods aside from `construct`.

Reply:
Thsnks! I was wondering the same. Good to know.